libraries/AC_Autorotation/AC_Autorotation.cpp

```diff
@@ -2,375 +2,524 @@
-
-    // Limiting the target velocity based on the max acceleration limit
-    if (_cmd_vel < _vel_target) {
-        _cmd_vel += _accel_max * _dt;
-        if (_cmd_vel > _vel_target) {
-            _cmd_vel = _vel_target;
-        }
-    } else {
-        _cmd_vel -= _accel_max * _dt;
-        if (_cmd_vel < _vel_target) {
-            _cmd_vel = _vel_target;
-        }
+    // enum class for bitmask documentation in logging
+    enum class AC_Autorotation_Landed_Reason : uint8_t {
+        LOW_SPEED = 1<<0, // true if below 1 m/s
+        LAND_COL  = 1<<1, // true if collective below land col min
+        IS_STILL  = 1<<2, // passes inertial nav is_still() check
+    };
+
+    uint8_t reason = 0;
+    if (_landed_reason.min_speed) {
+        reason |= uint8_t(AC_Autorotation_Landed_Reason::LOW_SPEED);
+    }
+    if (_landed_reason.land_col) {
+        reason |= uint8_t(AC_Autorotation_Landed_Reason::LAND_COL);
+    }
+    if (_landed_reason.is_still) {
+        reason |= uint8_t(AC_Autorotation_Landed_Reason::IS_STILL);
     }
 
-    // get p
-    _vel_p = _p_fw_vel.get_p(_cmd_vel-_speed_forward);
+    // @LoggerMessage: AROT
+    // @Vehicles: Copter
+    // @Description: Helicopter AutoROTation (AROT) information
+    // @Field: TimeUS: Time since system startup
+    // @Field: LR: Landed Reason state flags
+    // @FieldBitmaskEnum: LR: AC_Autorotation_Landed_Reason
 
-    // get ff
-    _vel_ff = _cmd_vel*_param_fwd_k_ff;
+    // Write to data flash log
+    AP::logger().WriteStreaming("AROT",
+                                "TimeUS,LR",
+                                "s-",
+                                "F-",
+                                "QB",
+                                AP_HAL::micros64(),
+                                _landed_reason);
+}
+#endif  // HAL_LOGGING_ENABLED
 
-    //calculate acceleration target based on PI controller
-    _accel_target = _vel_ff + _vel_p;
+// Arming checks for autorotation, mostly checking for miss-configurations
+bool AC_Autorotation::arming_checks(size_t buflen, char *buffer) const
+{
+    if (!enabled()) {
+        // Don't run arming checks if not enabled
+        return true;
+    }
+
+    const AP_HAL::HAL& hal = AP_HAL::get_HAL();
 
-    // filter correction acceleration
-    _accel_target_filter.set_cutoff_frequency(10.0f);
-    _accel_target_filter.apply(_accel_target, _dt);
+    // Check for correct RPM sensor config
+#if AP_RPM_ENABLED
+    // Get singleton for RPM library
+    const AP_RPM *rpm = AP_RPM::get_singleton();
 
-    //Limits the maximum change in pitch attitude based on acceleration
-    if (_accel_target > _accel_out_last + _accel_max) {
-        _accel_target = _accel_out_last + _accel_max;
-    } else if (_accel_target < _accel_out_last - _accel_max) {
-        _accel_target = _accel_out_last - _accel_max;
+    // Get current rpm, checking to ensure no nullptr
+    if (rpm == nullptr) {
+        hal.util->snprintf(buffer, buflen, "Can't access RPM");
```

Comment:
This case might be so bad that its not worth spending flash on the string. That is to say, if RPM is nullptr you have bigger issues.

Reply:
You do still need to check before using it of course, just we might be able to remove the warning.